.github/workflows/wheel.yaml

```diff
@@ -0,0 +1,106 @@
+name: Build and upload to PyPI
+
+on:
+  push:
+  pull_request:
+
+jobs:
+  build_wheels:
+    name: Build wheels for cp${{ matrix.python }}-${{ matrix.platform_id }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          # linux-64
+          - os: ubuntu-latest
+            python: 38
+            platform_id: manylinux_x86_64
+          - os: ubuntu-latest
+            python: 39
+            platform_id: manylinux_x86_64
+          - os: ubuntu-latest
+            python: 310
+            platform_id: manylinux_x86_64
+          - os: ubuntu-latest
+            python: 311
+            platform_id: manylinux_x86_64
+          # - os: ubuntu-latest
+          #   python: 312
+          #   platform_id: manylinux_x86_64
+          # macos-x86-64
+          - os: macos-latest
+            python: 38
+            platform_id: macosx_x86_64
+          - os: macos-latest
+            python: 39
+            platform_id: macosx_x86_64
+          - os: macos-latest
+            python: 310
+            platform_id: macosx_x86_64
+          - os: macos-latest
+            python: 311
+            platform_id: macosx_x86_64
+          #- os: macos-latest
+          #  python: 312
+          #  platform_id: macosx_x86_64
+          # win-64
+          # - os: windows-2019
+          #   python: 38
+          #   platform_id: win_amd64
+          # - os: windows-2019
+          #   python: 39
+          #   platform_id: win_amd64
+          # - os: windows-2019
+          #   python: 310
+          #   platform_id: win_amd64
+          # - os: windows-2019
+          #   python: 311
+          #   platform_id: win_amd64
+          # - os: windows-2019
+          #   python: 312
+          #   platform_id: win_amd64
+    steps:
+      - uses: actions/checkout@v4
+      - name: Build wheels
+        uses: pypa/cibuildwheel@v2.16
+        env:
+          CIBW_BUILD_VERBOSITY: 1
+          CIBW_ARCHS: all
+          CIBW_BUILD: cp${{ matrix.python }}-${{ matrix.platform_id }}
+      - uses: actions/upload-artifact@v3
+        with:
+          path: ./wheelhouse/*.whl
+  build_sdist:
+    name: Build source distribution
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - uses: actions/setup-python@v4
+        name: Install Python
+        with:
+          python-version: '3.12'
```

Comment:
This installs Python 3.12 in all builds - do you mean to use `matrix.python`?

Reply:
The `build_sdist` step only packages the source distribution, i.e. the file that ends with `.tar.gz`. The Python version here doesn't matter as all versions and platforms share the same source distribution.